### PR TITLE
fix: start/stop jobs when PATCHed [DHIS2-10890]

### PR DESCRIPTION
**Pull request by jbee**
Problem was that continues jobs could not be stopped. While one could disable it it would still run in the scheduler which after finishing a round would update the job configuration for the next due time and override the disabled state.

This would not be noticeable since the client began to use PATCH to flip the `enabled` flag. This would no longer trigger the `JobConfigurationObjectBundleHook` which had done this for POST.

I also noticed a thread-safety issue in the `DefaultSchedulingManager`.
It would do a `futures.contains(key)` followed by a `futures.put(key)`. Obviously the state could change between `contains` and `put` allowing for unintended things to happen when multiple threads are involved. This is now done via `computeIfAbsent` which makes sure only one thread will succeed and start/schedule a new task.

### CODE REVIEW

**Comment by larshelge** [comment]
Good catch, this looks great overall :+1:

**Review by larshelge — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/scheduling/DefaultSchedulingManager.java [hunk 1]
```diff
@@ -172,52 +173,48 @@ public void jobConfigurationFinished( JobConfiguration jobConfiguration )
...
+    }
 
-                log.info( String.format( "Scheduled job: %s", jobConfiguration ) );
-            }
-        }
+    private ScheduledFuture<?> scheduleCronOrDelayJob( JobConfiguration jobConfiguration, JobInstance jobInstance )
+    {
+        log.info( String.format( "Scheduling job: %s", jobConfiguration ) );
+        SchedulingType type = jobConfiguration.getJobType().getSchedulingType();
+        ScheduledFuture<?> future = type == SchedulingType.CRON
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/scheduling/DefaultSchedulingManager.java:183]: This is good. Just wondering, is readability suffering when we have such long lambdas/ternary statements, maybe make it slightly more old-school?

**Review by enricocolasante — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/scheduling/DefaultSchedulingManager.java (hunk 1 above)
Comment (jbee) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/scheduling/DefaultSchedulingManager.java:183]: I wasn't too happy about it either. Had tried a few variants. Enum with two values seems to throw off static analysis a bit too complaining about some solutions. Now looking at it it probably is a readable solution if both cases of the ternary get extracted to a helper method? WDYT?

Comment (jbee) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/scheduling/DefaultSchedulingManager.java]: Extracting made this readable.

**Pull request merged**